File: regular_practice/20220930/_走迷宫2.java
import java.util.*;
import java.io.*;

public class _走迷宫2 {
    static final int[][] DIRECTIONS = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    static int n, m;
    static char[][] g;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer in = new StreamTokenizer(br);
        in.nextToken();
        n = (int) in.nval;
        in.nextToken();
        m = (int) in.nval;
        g = new char[n][m];
        for (int i = 0; i < n; ++i) {
            g[i] = br.readLine().toCharArray();
        }
        System.out.println(bfs());
    }

    static int bfs() {
        boolean[][] vis = new boolean[n][m];
        Queue<int[]> que = new LinkedList<>();
        que.offer(new int[]{0, 0, 1});
        vis[0][0] = true;
        while (!que.isEmpty()) {
            int[] curr = que.poll();
            int x = curr[0], y = curr[1], step = curr[2];
            if (x == n - 1 && y == m - 1) return step;
            for (int[] d : DIRECTIONS) {
                int nx = x + d[0], ny = y + d[1];
                if (nx < 0 || nx >= n || ny < 0 || ny >= m
                    || g[nx][ny] == '#' || vis[nx][ny]) continue;
                que.offer(new int[]{nx, ny, step + 1});
                vis[nx][ny] = true;
            }
        }
        return -1;
    }
}

File: regular_practice/20220930/_走迷宫2.cpp
#include <iostream>
#include <limits>
#include <queue>

using namespace std;

const int MAX_N = 41;
const int DIRECTIONS[][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
int r, c;
char g[MAX_N][MAX_N];
bool vis[MAX_N][MAX_N];

struct State {
    int x, y, step;
};

int bfs() {
    queue<State> que;
    que.push({0, 0, 1});
    vis[0][0] = true;
    while (!que.empty()) {
        State state = que.front();
        que.pop();
        if (state.x == r - 1 && state.y == c - 1) return state.step;
        for (auto d : DIRECTIONS) {
            int nx = state.x + d[0], ny = state.y + d[1];
            if (nx < 0 || nx >= r || ny < 0 || ny >= c ||
                g[nx][ny] == '#' || vis[nx][ny]) continue;
            que.push({nx, ny, state.step + 1});
            vis[nx][ny] = true;
        }
    }

    return -1;
}

int main() {
    scanf("%d%d", &r, &c);
    cin.ignore(numeric_limits<streamsize>::max(), '\n');
    for (int i = 0; i < r; ++i) gets(g[i]);
    printf("%d", bfs());

    return 0;
}